src/test/java/io/numaproj/numaflow/function/FunctionServerTest.java

```diff
@@ -152,4 +152,45 @@ public void reducer() throws InterruptedException {
 
     }
 
+    @Test
+    public void reducerWithMultipleKey() {
+        String reduceKey = "reduce-key";
+        int keyCount = 100;
+        Udfunction.Datum.Builder inDatumBuilder = Udfunction.Datum.newBuilder().setKey(reduceKey);
+
+        Metadata metadata = new Metadata();
+        metadata.put(Metadata.Key.of(DATUM_KEY, Metadata.ASCII_STRING_MARSHALLER), reduceKey);
+        metadata.put(Metadata.Key.of(WIN_START_KEY, Metadata.ASCII_STRING_MARSHALLER), "60000");
+        metadata.put(Metadata.Key.of(WIN_END_KEY, Metadata.ASCII_STRING_MARSHALLER), "120000");
+
+        //create an output stream observer
+        ReduceOutputStreamObserver outputStreamObserver = new ReduceOutputStreamObserver();
+
+        StreamObserver<Udfunction.Datum> inputStreamObserver = UserDefinedFunctionGrpc
+                .newStub(inProcessChannel)
+                .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
+                .reduceFn(outputStreamObserver);
+
+        // send messages with 100 different keys
+        for (int j = 0; j < keyCount; j++) {
+            for (int i = 1; i <= 10; i++) {
+                Udfunction.Datum inputDatum = inDatumBuilder
+                        .setKey(reduceKey + j)
+                        .setValue(ByteString.copyFromUtf8(String.valueOf(i)))
+                        .build();
+                inputStreamObserver.onNext(inputDatum);
+            }
+        }
+
+        inputStreamObserver.onCompleted();
+
+        String expectedKey = reduceKey + reduceProcessedKeySuffix;
+        // sum of first 10 numbers 1 to 10 -> 55
+        ByteString expectedValue = ByteString.copyFromUtf8(String.valueOf(55));
+
+        Udfunction.DatumList result = outputStreamObserver.getResultDatum();
+        assertEquals(100, result.getElementsCount());
+        assertEquals(expectedValue, result.getElements(0).getValue());
```

Comment:
Should we assertEquals on all keys instead of only the first one?

Reply:
done